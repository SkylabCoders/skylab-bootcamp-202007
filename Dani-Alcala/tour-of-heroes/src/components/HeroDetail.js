import React, { useEffect, useState } from 'react';
import { Prompt } from 'react-router-dom';
import heroStore from '../stores/heroStore';
import { loadHeroes, saveHero } from '../actions/heroActions';
import TextInput from './TextInput';

function HeroDetail(props) {
	const [heroes, setHeroes] = useState(heroStore.getHeroes());
	const [heroId, setHeroId] = useState(null);
	const [heroName, setHeroName] = useState('');
	const [isFormDirty, setIsFormDirty] = useState(false);

	useEffect(() => {
		heroStore.addChangeListener(onChange);
		const heroId = +props.match.params.heroId;
		if (heroes.length === 0) {
			loadHeroes();
		} else if (heroId) {
			const hero = heroStore.getHeroById(heroId);
			if (hero) {
				setHeroName(hero.name);
				setHeroId(hero.id);
			}
		}
		return () => heroStore.removeChangeListener(onChange);
	}, [heroes.length, props.match.params.heroId]);

	function onChange() {
		setHeroes(heroStore.getHeroes());
	}

	function onFieldChange(value, setValue) {
		setValue(value);
		setIsFormDirty(true);
	}

	function handleSubmit(event) {
		event.preventDefault();
		saveHero({ name: heroName, id: heroId }).then(() =>
			props.history.push('/heroes')
		);
	}

	return (
		<form onSubmit={handleSubmit}>
			{!heroId && <h2>Register a new hero:</h2>}
			{heroId && (
				<>
					<h2>{heroName} details!</h2>
					<p>Id: {heroId}</p>
				</>
			)}
			<TextInput
				name="heroName"
				value={heroName}
				placeholder="Hero name"
				onChange={(event) => onFieldChange(event.target.value, setHeroName)}
			/>
			<Prompt
				when={isFormDirty}
				message="Are you sure you want to navigate away?"
			/>
			<p>
				<button type=" bv">Save Hero</button>
			</p>
		</form>
	);
}

export default HeroDetail;
