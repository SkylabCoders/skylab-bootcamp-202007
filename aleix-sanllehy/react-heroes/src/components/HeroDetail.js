import React from 'react';
import heroList from '../hero.mock';
import { Link, Prompt } from 'react-router-dom';

class HeroDetail extends React.Component {
	constructor(props) {
		super(props);
		//this.state = this.getId();
		//this.state = this.getHeroById(+this.props.match.params.heroId);
		this.state = { heroId: null, heroName: '', formIsDirty: false };
		this.onFieldChange = this.onFieldChange.bind(this);
	}

	onFieldChange(myEvent) {
		this.setState({ heroName: myEvent.target.value, formIsDirty: true });
		console.log(this.formIsDirty);
		//this.setState({ [myEvent.target.name]: myEvent.target.value });
		//console.log(myEvent.target.name, '=>', myEvent.target.value);
	}

	/* 	getId() {
		//const urlId = +window.location.pathnamesplit('/')[2]; // different approaches to get the hero id
		//const urlId = +this.props.location.pathname.split('/')[2];
		const urlId = +this.props.match.params.heroId;
		return heroData.find((element) => element.id === urlId);
	} */

	getHeroById(id) {
		return heroList.find((hero) => hero.id === id);
	}

	componentDidMount() {
		const hero = this.getHeroById(+this.props.match.params.heroId);
		this.setState({
			heroId: hero.id,
			heroName: hero.name
		});
	}

	render() {
		return (
			<main className="hero-details">
				<h3>
					<span>{this.state.heroName}</span> details!
				</h3>
				<form>
					<p className="hero-details__element">
						<span className="hero-details__id">id:</span>
						<span className="hero-details__id">{this.state.heroId}</span>
					</p>
					<label className="hero-details__name" htmlFor="heroName">
						name:
						<input
							name="heroName"
							placeholder="Hero name"
							value={this.state.heroName}
							onChange={this.onFieldChange}
						/>
					</label>
					<Prompt
						when={this.state.formIsDirty}
						message="Discard changes and go back?"
					/>
				</form>

				<Link className="button-back hero-element" to="/hero">
					Back
				</Link>
			</main>
		);
	}
}

export default HeroDetail;
