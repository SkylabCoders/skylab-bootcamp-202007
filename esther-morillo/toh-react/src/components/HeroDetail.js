import React, { useEffect, useState } from 'react';
import { Prompt } from 'react-router-dom';
import heroStore from '../stores/heroStore';
import { loadHeroes, saveHero } from '../actions/heroActions';
import TextInput from './TextInput';

function HeroDetail(props) {
	const [heroes, setHeroes] = useState(heroStore.getHeroes());
	const [heroId, setHeroId] = useState(null);
	const [heroName, setHeroName] = useState('');
	const [isFormDirty, setIsFormDirty] = useState(false);

	useEffect(() => {
		heroStore.addChangeListener(onChange);
		const heroId = +props.match.params.heroId;
		if (heroes.length === 0) {
			loadHeroes();
		} else if (heroId) {
			const hero = heroStore.getHeroById(heroId);
			if (hero) {
				setHeroName(hero.name);
				setHeroId(hero.id);
			}
		}
		return () => heroStore.removeChangeListener(onChange);
	}, [heroes.length, props.match.params.heroId]); //Esto es un array de condiciones. Si se cumple uno u otro vuelve a ejecutar. Si es un array vacío se ejecuta solo una vez

	function onChange() {
		setHeroes(heroStore.getHeroes());
	}

	function onFieldChange(value, setValue) {
		setValue(value);
		setIsFormDirty(true);
	}

	function handleSubmit(event) {
		event.preventDefault();
		//para que no coja espacios o que alguien escriba un espacio y ya lo coja como nombre
		const name = heroName.trim();
		
		if(name) {
			saveHero({ name, id: heroId }).then(() =>
				props.history.push('/heroes')
			);
		} else {
			alert('hero name is empty')
		}
	}

	return (
		<form onSubmit={handleSubmit}>
			{/* Si no hay un ID muestro el h2. Solo funciona ocn booleano delante (truco). No es decir esta condición y esta otra */}
			{!heroId && <h2>Register a new hero:</h2>}
			{/* SI hay un ID haz esto */}
			{heroId && (
				<>
					<h2>{heroName} details!</h2>
					<p>Id: {heroId}</p>
				</>
			)}
			<TextInput
				name="heroName"
				value={heroName}
				placeholder="Hero name"
				onChange={(event) => onFieldChange(event.target.value, setHeroName)}
			/>
			<Prompt
				when={isFormDirty}
				message="Are you sure you want to navigate away?"
			/>
			<p>
				<button type="submit">Save Hero</button>
			</p>
		</form>
	);
}

export default HeroDetail;
