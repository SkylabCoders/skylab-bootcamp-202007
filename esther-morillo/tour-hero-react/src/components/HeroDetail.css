* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

body {
    font-family: sans-serif;
}

.title {
    color: #969696;
    margin-bottom: 1rem;
}

.form__container {
    display: flex;
    justify-content: center;
}

.form {
    display: flex;
    flex-direction: column;
}

.detail__title {
    color: #404040;
    margin: 1.5rem 0;
}

.detail__name,
.inpup__name {
    color: #557584;
    font-weight: bold;
}

.id-name span {
    color: #969696;
    font-weight: lighter;
    padding-left: 1.8rem;
}

.input {
    margin-top: .8rem;
    margin-left: .5rem;
    padding: .7rem;
    border: .09rem solid rgb(209, 208, 208);
    border-radius: .2rem;
}

.input:focus {
    outline: 0 none;
}

.button-back {
    background-color: #eee;
    color: #969696;
    width: 4rem;
    margin-right: .5rem;
    padding: .5rem 1rem;
    border-radius: .3rem;
    margin-top: 1rem;
    cursor: pointer;
}