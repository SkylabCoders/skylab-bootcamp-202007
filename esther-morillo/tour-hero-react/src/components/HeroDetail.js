import React from 'react';
import './heroDetail.css';
import heroList from '../hero-service';
import { Prompt } from 'react-router-dom';


class HeroDetail extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			heroName: '',
			heroId: null,
			forIsDirty: false
		};
		this.onFieldChange = this.onFieldChange.bind(this);
	}

	onFieldChange(myEvent) {
		this.setState({
			[myEvent.target.name]: myEvent.target.value,
			forIsDirty: true
		});
	}

	getHeroId(id) {
		//Me devuelve el objeto entero que coincida con el id que pasamos (pathHeroId)
		return heroList.find((hero) => hero.id === id);
	}

	componentDidMount() {
		//Â¡Cuidado! Que el id de la url viene como string (+)
		// heroId se lo hemos puesto nosotros en App.js ==> :heroId 
		// console.log(this.props.match.params) ==> me devuelve => {heroId: "16"} 
		// this.props.match.params.heroId => Me devuelve el id
		console.log(this.props.match.params)
		const hero = this.getHeroId(+this.props.match.params.heroId);
		//Actualizamos state
		this.setState({
			heroName: hero.name,
			heroId: hero.id
		})
	}


	render() {
		return (
			<div className="form__container">
			<form className="form">
				<h2 className="detail__title">{this.state.heroName} details!</h2>
				<p className="detail__name">id: {this.state.heroId}</p>
				<label className="detail__name" htmlFor="heroName">
					name:
					<input
						className="input"
						name="heroName"
						placeholder="Hero name"
						value={this.state.heroName}
						onChange={this.onFieldChange}
					/>
				</label>

				<Prompt 
					when={this.state.forIsDirty}
					message="Are you sure you want to navigate away?"
				/>

			</form>
			</div>
		);
	}
}

export default HeroDetail;
