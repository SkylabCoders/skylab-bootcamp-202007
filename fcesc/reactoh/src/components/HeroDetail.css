.HeroDetail{
    text-align: center;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    align-items: center;
}