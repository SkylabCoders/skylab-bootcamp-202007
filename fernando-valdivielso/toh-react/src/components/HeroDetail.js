import React, { useEffect, useState } from 'react';
import { Prompt } from 'react-router-dom';
import heroStore from '../stores/heroStore';
import { loadHeroes, saveHero } from '../actions/heroActions';
import TextInput from './TextInput';

function HeroDetail(props) {
	const [heroes, setHeroes] = useState(heroStore.getHeroes());
	const [heroId, setHeroId] = useState(null);
	const [heroName, setHeroName] = useState('');
	const [isFormDirty, setIsFormDirty] = useState(false);

	useEffect(() => {
		heroStore.addChangeListener(onChange);
		const heroId = +props.match.params.heroId;
		if (heroes.length === 0) {
			loadHeroes();
		} else if (heroId) {
			const hero = heroStore.getHeroById(heroId);
			if (hero) {
				setHeroName(hero.name);
				setHeroId(hero.id);
			}
		}
		return () => heroStore.removeChangeListener(onChange);
	}, [heroes.length, props.match.params.heroId]);

	function onChange() {
		setHeroes(heroStore.getHeroes());
	}

	function onFieldChange(value, setValue) {
		setValue(value);
		setIsFormDirty(true);
	}

	function handleSubmit(event) {
		event.preventDefault();
		saveHero({ name: heroName, id: heroId }).then(() =>
			props.history.push('/heroes')
		);
	}

	return (
		<form onSubmit={handleSubmit}>
			{!heroId && <h2>Register a new hero:</h2>}
			{heroId && (
				<>
					<h2>{heroName} details!</h2>
					<p>Id: {heroId}</p>
				</>
			)}
			<TextInput
				name="heroName"
				value={heroName}
				placeholder="Hero name"
				onChange={(event) => onFieldChange(event.target.value, setHeroName)}
			/>
			<Prompt
				when={isFormDirty}
				message="Are you sure you want to navigate away?"
			/>
			<p>
				<button type="submit">Save Hero</button>
			</p>
		</form>
	);
}

export default HeroDetail;












// class HeroDetail extends React.Component {
// constructor(props) {
// 	super(props);

// 	this.state = {
// 		heroName: '',
// 		heroId: '',
// 		formIsDirty: false
// 	};

// 	this.onFieldChange = this.onFieldChange.bind(this);
// }


// componentDidMount() {

// 	let heroe = this.findHeroById(+this.props.match.params.heroId)
// 	this.setState({
// 		heroName: heroe.name,
// 		heroId: heroe.id
// 	})
// }

// findHeroById(heroId) {
// 	return heroData.find((hero) => hero.id === heroId)

// }

// onFieldChange(myEvent) {
// 	this.setState({
// 		[myEvent.target.name]: myEvent.target.value, //[myEvent.target.name] = heroName 
// 		formIsDirty: true

// 	});

// }


// render() {
// 	return (
// 		<form>
// 			<p>{this.state.heroName} details!</p>
// 			<p>id: {this.state.heroId}</p>
// 			<label htmlFor="heroName">

// 				name:
// 					<input
// 					name="heroName"
// 					placeholder="Hero name"
// 					value={this.state.heroName}
// 					onChange={this.onFieldChange}
// 				/>

// 			</label>
// 			<button>Create Hero</button>
// 			<button>Update Hero</button>
// 			<Prompt
// 				when={this.state.formIsDirty}
// 				message='Are you sure you wamt to exit this page?'
// 			/>
// 		</form>
// 	);
// }
// }

// export default HeroDetail;
