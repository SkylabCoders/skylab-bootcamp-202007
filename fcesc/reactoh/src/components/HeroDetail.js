import React, {useState, useEffect} from 'react';
import './HeroDetail.css';
import DisplayInlineArray from './DisplayInlineArray';
import ShowImage from './ShowImage';
import HeroStore from './../stores/heroStore';
import { NavLink, useRouteMatch } from 'react-router-dom';
import { loadHeroById } from './../actions/heroActions';

function HeroDetail(){
    let urlQuery = useRouteMatch()
    let urlId = +urlQuery.params.heroId;
    const [currentHero, setcurrentHero] = useState(undefined);
    const [id] = useState(urlId);

    useEffect(()=>{
        HeroStore.addChangeListener(onChange);
        if( currentHero === undefined){ loadHeroById(id) }
        return ()=>{HeroStore.removeChangeListener(onChange);}
    }, []);

    function onChange(){
        setcurrentHero(HeroStore.getHeroById(id)); 
    }

    console.log(currentHero);

    return (        
        <div className="heroDetail">
            <h2>This is a test</h2>
            <div className="heroDetail__image">
                <ShowImage
                    imgUrl={currentHero.images.md}
                />
            </div>
            <div className="heroDetail__data">
                <h2>{currentHero.name}</h2>
                <p>Id: {currentHero.id}, real name: <strong>{currentHero.biography.fullName}</strong>, alter egos: {currentHero.biography.alterEgos}</p>
                <p>Place of birth: {currentHero.biography.placeOfBirth}, alias: <DisplayInlineArray aliases={currentHero.biography.aliases}/></p>
                <p>Race: {currentHero.appearance.race}, gender: {currentHero.appearance.gender}, height: {currentHero.appearance.height[1]}, weight: {currentHero.appearance.weight[1]}</p>
                <p>Occupation: {currentHero.work.occupation}, alignment: {currentHero.biography.alignment}</p>
                <p>First appearance: {currentHero.biography.firstAppearance}, publisher {currentHero.biography.publisher}</p>
                <NavLink to="/paginated-flux">Go back to list</NavLink>
            </div>
        </div>
    );
}

export default HeroDetail;