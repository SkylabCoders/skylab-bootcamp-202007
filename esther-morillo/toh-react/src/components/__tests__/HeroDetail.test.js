import renderer from 'react-test-renderer';
import React from 'react'
import HeroDetail from '../HeroDetail';
import { BrowserRouter } from 'react-router-dom';


describe('Detail snapshot', () => {
    const props = {
        match: {
            params: {
                heroId: 14
            }
        }
    };
    const heroDetailTree = renderer.create(
    <BrowserRouter>
        <HeroDetail {...props}/>
    </BrowserRouter>
    );
    
    it('should match', () => {
        expect(heroDetailTree).toMatchSnapshot();
    });  
})