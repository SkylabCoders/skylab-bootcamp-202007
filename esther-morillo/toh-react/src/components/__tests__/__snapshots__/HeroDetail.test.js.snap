// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail snapshot should match 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Register a new hero:
  </h2>
  <label
    htmlFor="heroName"
  >
    Name: 
    <input
      name="heroName"
      onChange={[Function]}
      placeholder="Hero name"
      type="text"
      value=""
    />
  </label>
  <p>
    <button
      type="submit"
    >
      Save Hero
    </button>
  </p>
</form>
`;
